File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInsidePresentationControllersAdaptivityandCustomAnimatorObjects/LookInside/AAPLOverlayPresentationController.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  AAPLOverlayPresentationController implementation.
  
 */

#import "AAPLOverlayPresentationController.h"
#import "AAPLOverlayTransitioner.h"

@implementation AAPLOverlayPresentationController

- (instancetype)initWithPresentingViewController:(UIViewController *)presentingViewController presentedViewController:(UIViewController *)presentedViewController
{
    self = [super initWithPresentedViewController:presentedViewController presentingViewController:presentingViewController];
    if(self)
    {
        [self prepareDimmingView];
    }
    
    return self;
}

- (void)presentationTransitionWillBegin
{
    UIView* containerView = [self containerView];
    UIViewController* presentedViewController = [self presentedViewController];
    [dimmingView setFrame:[containerView bounds]];
    [dimmingView setAlpha:0.0];

    [containerView insertSubview:dimmingView atIndex:0];
    
    if([presentedViewController transitionCoordinator])
    {
        [[presentedViewController transitionCoordinator] animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
            [dimmingView setAlpha:1.0];
        } completion:nil];
    }
    else
    {
        [dimmingView setAlpha:1.0];
    }
}

- (void)dismissalTransitionWillBegin
{
    if([[self presentedViewController] transitionCoordinator])
    {
        [[[self presentedViewController] transitionCoordinator] animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
            [dimmingView setAlpha:0.0];
        } completion:nil];
    }
    else
    {
        [dimmingView setAlpha:0.0];
    }
}

- (UIModalPresentationStyle)adaptivePresentationStyle
{
    return UIModalPresentationOverFullScreen;
}

- (CGSize)sizeForChildContentContainer:(id <UIContentContainer>)container withParentContainerSize:(CGSize)parentSize
{
    return CGSizeMake(floorf(parentSize.width / 3.0),
                      parentSize.height);
}

- (void)containerViewWillLayoutSubviews
{
    [dimmingView setFrame:[[self containerView] bounds]];
    [[self presentedView] setFrame:[self frameOfPresentedViewInContainerView]];
}

- (BOOL)shouldPresentInFullscreen
{
    return YES;
}

- (CGRect)frameOfPresentedViewInContainerView
{
    CGRect presentedViewFrame = CGRectZero;
    CGRect containerBounds = [[self containerView] bounds];
    
    presentedViewFrame.size = [self sizeForChildContentContainer:(UIViewController<UIContentContainer> *)[self presentedViewController]
                                         withParentContainerSize:containerBounds.size];
    
    presentedViewFrame.origin.x = containerBounds.size.width - presentedViewFrame.size.width;
    
    return presentedViewFrame;
    
}

@synthesize dimmingView;

- (void)prepareDimmingView
{
    dimmingView = [[UIView alloc] init];
    [dimmingView setBackgroundColor:[UIColor colorWithWhite:0.0 alpha:0.4]];
    [dimmingView setAlpha:0.0];
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dimmingViewTapped:)];
    [dimmingView addGestureRecognizer:tap];
}

- (void)dimmingViewTapped:(UIGestureRecognizer *)gesture
{
    if([gesture state] == UIGestureRecognizerStateRecognized)
    {
        [[self presentingViewController] dismissViewControllerAnimated:YES completion:NULL];
    }
}

@end

File: UIKit and Animations/WWDC 2014/View Controller Advancements for iOS8/Sample Codes/CustomPresentation/CustomPresentation/AAPLOverlayPresentationController.swift
//
//  AAPLOverlayPresentationController.swift
//  LookInside_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/18/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class AAPLOverlayPresentationController: UIPresentationController {
	
	var dimmingView: UIView!
	var size: CGSize?
	
	override init(presentedViewController: UIViewController, presenting presentingViewController: UIViewController?) {
		super.init(presentedViewController: presentedViewController, presenting: presentingViewController)
		self.prepareDimmingView()
	}
	
	override func presentationTransitionWillBegin() {
		let containerView = self.containerView
		let presentedViewController = self.presentedViewController
		self.dimmingView.frame = containerView!.bounds
		self.dimmingView.alpha = 0.0
		
		containerView?.insertSubview(self.dimmingView, at: 0)
		
		if presentedViewController.transitionCoordinator != nil {
			presentedViewController.transitionCoordinator?.animate(alongsideTransition: { (context) in
				self.dimmingView.alpha = 1.0
			}, completion: nil)
		}
		else {
			self.dimmingView.alpha = 1.0
		}
	}
	
	override func dismissalTransitionWillBegin() {
		let presentedViewController = self.presentedViewController

		if presentedViewController.transitionCoordinator != nil {
			presentedViewController.transitionCoordinator?.animate(alongsideTransition: { (context) in
				self.dimmingView.alpha = 0.0
			}, completion: nil)
		}
		else {
			self.dimmingView.alpha = 0.0
		}
	}
	
	override var adaptivePresentationStyle: UIModalPresentationStyle {
		return .overFullScreen
	}
	
	override func size(forChildContentContainer container: UIContentContainer, withParentContainerSize parentSize: CGSize) -> CGSize {
		let width = floor(parentSize.width * 0.9)
		let height = floor(parentSize.height * 0.7)

		return self.size ?? CGSize(width: width, height: height)
	}
	
	override func containerViewWillLayoutSubviews() {
		self.dimmingView.frame = self.containerView!.bounds
		self.presentedView?.frame = self.frameOfPresentedViewInContainerView
	}
	
	override var shouldPresentInFullscreen: Bool {
		return true
	}
	
	override var frameOfPresentedViewInContainerView: CGRect {
		var presentedViewFrame = CGRect.zero
		let containerBounds = self.containerView!.bounds
		presentedViewFrame.size = self.size(forChildContentContainer: self.presentedViewController, withParentContainerSize: containerBounds.size)
		presentedViewFrame.origin.x = (containerBounds.size.width - presentedViewFrame.size.width)/2
		presentedViewFrame.origin.y = (containerBounds.size.height - presentedViewFrame.size.height)/2

		return presentedViewFrame
	}
	
	func prepareDimmingView() {
		self.dimmingView = UIView()
		self.dimmingView.backgroundColor = UIColor(white: 0.0, alpha: 0.4)
		self.dimmingView.alpha = 0.0
		
		let tap = UITapGestureRecognizer(target: self, action: #selector(dimmingViewTapped(_:)))
		self.dimmingView.addGestureRecognizer(tap)
	}
	
	@objc func dimmingViewTapped(_ gesture: UIGestureRecognizer) {
		if gesture.state == .recognized {
			self.presentedViewController.dismiss(animated: true, completion: nil)
		}
	}
}
